fixes #14618 - Introduced configurable initialization of modules and providers

dmitri-d opened the pull request.

dmitri-d commented:
Work in progress, although most functionality has been added. 

This adds functionality to customize/override configuration and initialization of plugins and providers. 
- ability to update/add computed settings (as opposed to purely config-file based settings).
- ability to add and execute custom validators
- ability to start background services during plugin initialization

Configuration/initialization logic has been moved to https://github.com/theforeman/smart-proxy/pull/390/files#diff-11cebac8a6625c7bb985371a365a48e3R163 and https://github.com/theforeman/smart-proxy/pull/390/files#diff-11cebac8a6625c7bb985371a365a48e3R196. 

Plugin dsl adds support for overriding the classes above, but also provides ability to individually override validator, dependency, dependency injection wirings loader classes. 

Some questions that need answers:
- I noticed that one of the common issues is referring to wrong plugin class when retrieving provider settings. I'd like to merge main plugin and provider settings (per provider).
- Right now all runtime dependencies (think a bunch of `require`s that used to go in the `after_activation` block) are loaded before validators are being executed (to load custom validators classes). Instead, validator loader can perform that, which would allow us to delay loading of classes until after we know the configuration is valid.

Deep dive, examples, and docs are coming. I'll be happy to do a code walk through to help with the review(s). 

@ekohl, @helge000, @GregSutcliffe, @domcleal: thoughts?

ekohl commented:
It would help me if I saw what would change for me as a plugin author because it's hard for me to derive the changes from this diff.

dmitri-d commented:
I will port https://github.com/theforeman/smart-proxy/pull/376 PR to highlight the changes.

dmitri-d commented:
[test]

dmitri-d committed:
Fixes #14618 - refactored module initialization

  to allow for:
    - support for multiple simultaneous providers
    - programmatically defined settings
    - custom validators
    - ability to customize module loading

  Also includes support for constructor-based dependency injection

dmitri-d commented:
@ekohl: pls see https://github.com/theforeman/smart-proxy/pull/406 that demonstrates most of the changes in this PR.

dmitri-d commented:
I'm aware of build failures under 1.8.7 and am looking into them.

domcleal commented on lib/proxy/pluggable.rb:
```diff
@@ -1,131 +1,137 @@
+  attr_reader :plugin_name, :version, :after_activation_blk, :class_loader
+
+  # Methods below define DSL for defining plugins
+  def after_activation(&blk)
```
Is this method meant to be deprecated?  There are references in the loading code that it's "legacy".

dmitri-d replied:
Not at the moment, although it can be deprecated if I add support for blocks as means of loading classes, specifying di wiring, etc, all of which require class(es) atm.

domcleal commented on lib/proxy/pluggable.rb:
```diff
@@ -1,131 +1,137 @@
-    def version
-      self.class.version
-    end
+  def bundler_group(name)
```
missing test coverage?

domcleal replied:
Seems to be fixed at https://github.com/theforeman/smart-proxy/pull/390/files#diff-192cbec985fe34a2e50b9569df0975f4R43

dmitri-d replied:
I added a few tests, could one of those...

domcleal commented on lib/proxy/plugin_initializer.rb:
```diff
@@ -1,36 +1,287 @@
-  def instantiate_plugins(loaded_plugins)
-    loaded_plugins.dup.map {|p| p.update(:instance => p[:class].new)}
+  def notify_group_about_failure
+    #providers.each {|p| p.stop }
```
commented code - this should be implemented properly, or removed

domcleal commented on lib/proxy/settings.rb:
```diff
@@ -20,4 +20,9 @@ def self.load_plugin_settings(defaults, settings_file, settings_directory = nil)
+    YAML.load(File.read(File.join(settings_directory || ::Proxy::SETTINGS.settings_directory, settings_file))) || {}
+  end
 end
+
```
Nitpick, stray line.

domcleal commented on lib/smart_proxy_main.rb:
```diff
@@ -9,20 +9,22 @@
 require 'proxy/settings/plugin'
 require 'proxy/settings/global'
 require 'proxy/dependency_injection'
+require 'proxy/default_di_wirings'
```
Perhaps require this only from lib/proxy/plugin_initializer.rb, which is where it's used, rather than at the app level.  Probably the same for default plugin validators?  (Saves also doing it in the tests.)

domcleal commented on modules/dhcp_isc/dhcp_isc_main.rb:
```diff
@@ -5,25 +5,15 @@ module Proxy::DHCP::ISC
-      super(Proxy::DhcpPlugin.settings.server)
-      @config_file = Proxy::DHCP::ISC::Plugin.settings.config
-      @leases_file = Proxy::DHCP::ISC::Plugin.settings.leases
+    def initialize(provider_settings = OpenStruct.new, main_settings = OpenStruct.new, service = nil)
```
Why is only this class refactored in this way?  Perhaps leave it for now, and do them all together?

dmitri-d replied:
I cannot remember anymore, probably because I used that for testing. There's a PR for dhcp changes: https://github.com/theforeman/smart-proxy/pull/409.

domcleal replied:
Okay, can you revert it for now please?

domcleal commented on lib/proxy/plugin.rb:
```diff
@@ -77,28 +19,38 @@ def find_provider(provider_name)
 #  requires :another_plugin, "~> 1.3.0"
 #  default_settings :first => 'first', :second => 'second'
+#
+# load_classes 'a/b', 'a/c', ...
```
Nitpick, indentation is out - should match above+below.

domcleal commented:
1) The changes to the public plugin API are difficult to understand, particularly the DI terminology and to understand how the new APIs are used. This would be a good time to improve the documentation, both to help me review and for future use.

The wiki page would be a good start, provided it has a warning to say that this is >= 1.12 only.  I'm also quite concerned to understand which APIs (if any) are changing behaviour in this patch, are being deprecated, or are being removed.

dmitri-d commented:
The docs are in plans for the next week. I would suggest taking a look at https://github.com/theforeman/smart-proxy/pull/406 to see how plugins can use the new functionality (puppet_proxy_legacy module is probably one of the better examples).

dmitri-d commented:
Done with the first pass at the docs: http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Smart-Proxy_Plugin

dmitri-d commented:
Rebased, fixed issues pointed out above (other than changes to dhcp_isc module). Also:
- ::Proxy::Pluggable.load_classes now accepts a block as a parameter
- DI wirings no longer use block parameter (as it was redundant)
- renamed ::Proxy::Pluggable.load_runtime_configuration to ::Proxy::Pluggable.load_programmable_settings (the former was too ambiguous)

>  I'm also quite concerned to understand which APIs (if any) are changing behaviour in this patch, are being deprecated, or are being removed.

No existing APIs have been removed, deprecated, or changed behaviour (or not intentionally).

domcleal commented on lib/proxy/pluggable.rb:
```diff
@@ -1,131 +1,129 @@
+  end
 
-    # Methods below define DSL for defining plugins
+  def load_programmable_settings(class_name_to_use)
```
Can this accept a block as well, in the same way as load_classes?

Or even only accept a block?  It seems needlessly complex to create and instantiate classes just to call a method in both cases.

dmitri-d replied:
While having a class is much heavier than passing in a block, I prefer the former as it's a) testable (or easier to test), b) doesn't pollute plugin class with configuration logic/executable code (this is the nastiest example I have, there are other cases that aren't as bad, but still: https://github.com/theforeman/smart-proxy/pull/406/commits/5e1b066e520f29e7d5b9dee1dfb1466f350ff261#diff-9b357977669579ae92d74c22420de796R7)

domcleal replied:
Ah I see, that makes more sense with the complexity of that method.  I wonder if keeping the block as an option's best for consistency, allowing either to be used.

Is there any reason to require a class (and then instantiate it) over just passing something that responds to the correct method, so it'd be a class method or pass an instance of the class?

dmitri-d replied:
I can add support for blocks in load_programmable_settings and load_dependency_injection_wirings methods, although I'm concerned a bit that those will be abused.

Not sure I understand you correctly -- how would you specify an instance of a class (or anything else other than a class) in these spots?

domcleal replied:
> Not sure I understand you correctly -- how would you specify an instance of a class (or anything else other than a class) in these spots?

Yeah, it's not possible now because `.new` is called.  I was trying to understand why that was needed, when this could just accept something that responds to `#load_programmable_settings` and not call `.new` first.  It's just the instantiation of the class seems like a peculiar interface in Ruby, where just responding to methods is usually enough to implement an interface.

Rather than `@programmable_settings.new.load_programmable_settings`, I'd expect `@programmable_settings.load_programmable_settings`, then either change `load_programmable_settings` to a class method or pass an instance of the Proxy::Example::RuntimeConfiguration in here.

domcleal commented on lib/proxy/pluggable.rb:
```diff
@@ -1,131 +1,129 @@
 
-  module ClassMethods
-    attr_reader :plugin_name, :version, :after_activation_blk, :bundler_group_name
+  def load_dependency_injection_wirings(class_name_to_use)
```
ditto, a block here might be simpler?

dmitri-d replied:
Similar reasoning to the above, and an example to go along: https://github.com/theforeman/smart-proxy/pull/406/commits/5e1b066e520f29e7d5b9dee1dfb1466f350ff261#diff-9b357977669579ae92d74c22420de796R53

domcleal commented on lib/proxy/pluggable.rb:
```diff
@@ -1,131 +1,129 @@
-      raise ::Proxy::PluginLoadingAborted, message
-    end
+  def load_validators(class_name_to_use)
+    @validator_loader_class = class_name_to_use
```
and ditto, a block or even just accept a hash of validator names to classes.  I don't see the need for the extra class, instantiation and method.

dmitri-d replied:
This would work, considering that validator classes can be loaded in load_classes or even directly.

lzap commented:
When you are done with review, let me know. I would like to test this patch with my plugin that has a background service. Until now, there was no way to properly stop it.

dmitri-d commented:
- added support for blocks in `::Proxy::Pluggable#load_classes`, `::Proxy::Pluggable#load_programmable_settings`, and `::Proxy::Pluggable#load_dependency_injection_wirings`
- `::Proxy::Pluggable#load_validators` now expects a map of validator names to classses
- `::Proxy::Pluggable#class_loader`, `::Proxy::Pluggable#programmable_settings`, and `::Proxy::Pluggable#di_wirings` now return an instance of a class/block wrapper
- reverted changes in `dhcp_isc/dhcp_isc_main.rb` and related tests

domcleal commented on lib/proxy/error.rb:
```diff
@@ -1,5 +1,12 @@
 module Proxy::Error
-  class BadRequest < StandardError; end
-  class Unauthorized < StandardError; end
+  class HttpError < StandardError
```
Is this related?  I can't see where it's used.

domcleal commented on lib/proxy/pluggable.rb:
```diff
@@ -1,131 +1,170 @@
-      settings.each { |setting| validators << ::Proxy::PluginValidators::Presence.new(self, setting) }
+  class SettingsProcWrapper
+    def initialize(a_blk); @a_blk = a_blk; end
+    def load_programmable_settings(settings); @a_blk.call(settings); end
```
Can you expand these out rather than using semicolons?  It's more readable - also in classes above and below.

domcleal commented on lib/proxy/pluggable.rb:
```diff
@@ -1,131 +1,170 @@
-    # End of DSL
+  class DiWiringsProcWrapper
+    def initialize(a_blk); @a_blk = a_blk; end
+    def load_dependency_injection_wirings(container, settings); @a_blk.call(container, settings); end
```
particularly here

domcleal commented on test/dhcp_isc/server_isc_test.rb:
```diff
@@ -56,8 +56,11 @@ def test_omcmd_server_connect
     @dhcp.omcmd('connect')
-    assert_equal "port 999", omio.input_commands[1]
-    assert_equal "server 192.168.122.1", omio.input_commands[0]
+    assert_equal 'key key_name "key_secret"', omio.input_commands[0]
```
Unrelated changes?

domcleal commented on test/dns_dnscmd/dnscmd_test.rb:
```diff
@@ -86,7 +86,7 @@ def test_dns_zone_matches_second_best_match_if_zone_name_equals_host_name
 
   def test_dns_zone_matches_sole_available_zone
-    assert_equal('sole.domain', Proxy::Dns::Dnscmd::Record.new.match_zone('host.foo.bar.sole.domain', ["sole.domain"]))
+    assert_equal('sole.domain', Proxy::Dns::Dnscmd::Record.new('localhost', 86_400).match_zone('host.foo.bar.sole.domain', ["sole.domain"]))
```
ditto, unrelated?

domcleal commented on test/plugins/presence_validator_test.rb:
```diff
@@ -1,26 +1,24 @@
 require 'test_helper'
+require 'ostruct'
```
Unused?

domcleal commented on test/plugins/plugin_test.rb:
```diff
@@ -0,0 +1,165 @@
+  def test_plugin_loads_http_rack_path
+    assert !TestPlugin12.http_enabled?
+    assert TestPlugin12.https_enabled?
+    assert TestPlugin12.plugin_name, 'test12'
```
`assert_equal 'test12', TestPlugin12.plugin_name`

domcleal commented on test/plugins/plugin_test.rb:
```diff
@@ -0,0 +1,165 @@
+    File.stubs(:read).returns("require 'test12/test12_api'")
+
+    assert_equal TestPlugin12.http_rackup, ''
+    assert_equal TestPlugin12.https_rackup, "require 'test12/test12_api'"
```
argument order's wrong: expected, actual

domcleal commented on lib/proxy/plugin_initializer.rb:
```diff
@@ -1,36 +1,288 @@
+    not_available = used_providers - providers.map {|p| p[:name].to_sym}
+    return @providers = providers.map {|p| p[:class]} if not_available.empty?
+
+    fail_group_with_message("Disabling all modules in the group '#{member_names}': following providers are not available #{not_available}")
```
`member_names` here should be the group name?  Or should it be a list of members - if so, it should probably be printed more clearly with a join(", ") or something, as it appears as:

```
Disabling all modules in the group '[:hello]' due 
```

domcleal commented on lib/proxy/plugin_initializer.rb:
```diff
@@ -1,36 +1,288 @@
+  end
+
+  def fail_group(an_exception)
+    fail_group_with_message("Disabling all modules in the group '#{member_names}' due to a failure in one of them: #{an_exception}", an_exception.backtrace)
```
ditto, re member_names/group name

dmitri-d commented:
Rebased, fixed the issues pointed out in the review.

domcleal commented on test/dhcp_isc/server_isc_test.rb:
```diff
@@ -40,7 +39,7 @@ def test_isc_provider_initialization
     ::Proxy::DHCP::ISC::Plugin.load_test_settings(:config => 'config_file', :leases => 'leases_file',
                                                   :omapi_port => '7777', :key_name => 'key_name',
-                                                  :key_secret => 'key_secret')
+                                                  :key_secret => 'key_secret', )
```
This breaks the test on Ruby 1.8.7, but like the previous comment, I don't think this file needs any changes.  Could you revert it to what's in develop please?

domcleal commented on test/dns_dnscmd/dnscmd_test.rb:
```diff
@@ -128,6 +128,6 @@ def test_dnscmd_enum_zones_parses_primary_zones_only
       "bar.domain.local",
       "domain.local",
-      "TrustAnchors"], Proxy::Dns::Dnscmd::Record.new.enum_zones
+      "TrustAnchors"], Proxy::Dns::Dnscmd::Record.new('localhost', 86_400).enum_zones
```
ditto, not sure this is related?

dmitri-d commented:
fixed the two issues above.

domcleal commented:
Would you mind updating the wiki page again to reflect any recent changes (e.g. support for blocks)?

domcleal commented:
Merged as 342eeda, thanks @witlessbird.

dmitri-d commented:
> Would you mind updating the wiki page again to reflect any recent changes (e.g. support for blocks)?

Done.
